src/VisualStudio/Core/Def/ProjectSystem/CPS/IWorkspaceProjectContextFactory.cs

```diff
@@ -42,5 +37,97 @@ internal interface IWorkspaceProjectContextFactory
             string? binOutputPath,
             string? assemblyName,
             CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Creates and initializes a new project and returns a <see
+        /// cref="IWorkspaceProjectContext"/> to lazily initialize the properties and items for the
+        /// project.  This method guarantees that either the project is added (and the returned task
+        /// completes) or cancellation is observed and no project is added.
+        /// </summary>
+        /// <param name="id">Project guid</param>
+        /// <param name="uniqueName">
+        /// A string that uniquely identifies the workspace within the solution, for the lifetime of that solution.
+        ///
+        /// The exact form of this string is an implementation detail and can change without notice, so the value
+        /// should not be parsed. However, the contents of the string might be helpful for debugging so including
+        /// it in logging is reasonable. It may contain PII.
+        ///
+        /// This identifier will be consistent regardless of project renames.
+        /// </param>
+        /// <param name="data">Provider access to msbuild evaluation data for the project.</param>
+        /// <param name="hostObject">The IVsHierarchy for the project; this is used to track linked files across multiple projects when determining contexts.</param>
+        /// <exception cref="InvalidOperationException">A required property or item is not present in <see cref="EvaluationData"/> or has invalid value.</exception>
+        Task<IWorkspaceProjectContext> CreateProjectContextAsync(Guid id, string uniqueName, EvaluationData data, object? hostObject, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Names of msbuild properties whose values that <see cref="CreateProjectContextAsync(Guid, string, EvaluationData, object?, CancellationToken)"/> will receive via <see cref="EvaluationData"/>.
+        /// </summary>
+        ImmutableArray<string> EvaluationPropertyNames { get; }
+
+        /// <summary>
+        /// Types of msbuild items whose values and metadata that <see cref="CreateProjectContextAsync(Guid, string, EvaluationData, object?, CancellationToken)"/> will receive via <see cref="EvaluationData"/>.
+        /// </summary>
+        ImmutableArray<string> EvaluationItemTypes { get; }
     }
+
+    internal abstract class EvaluationData
```

Comment:
I think an abstract class is preferable as it can be updated without dual insertion and it can also define non-abstract methods that are part of the semantic contract.